### ISSUE TITLE
non-matching lineage table and treeview for certain peptides

### ISSUE THREAD

**Issue by bmesuere** [unipept/unipept#443]
For these two sequencues (possibly more, only tested sample7) `TNTILQSAFFR` and `MIIGGGSAYSR`, the `firmicutes` are listed as a child of the `bacteria` in the lineage table, but according to the tree they are a child from `Bacteroidetes/Chlorobi group` which is also incorrectly calculated as their LCA.

See http://unipept.ugent.be/sequences/TNTILQSAFFR/equateIL or the following screenshots:

![screen shot 2014-12-10 at 17 51 17](https://github.ugent.be/github-enterprise-assets/0000/0022/0000/0268/cba110ee-808c-11e4-9b0e-0568bfe9c97a.png)
![screen shot 2014-12-10 at 17 51 23](https://github.ugent.be/github-enterprise-assets/0000/0022/0000/0269/ce71ae6e-808c-11e4-8630-d2fa13f1d317.png)

_[Original issue](https://github.ugent.be/unipept/unipept/issues/443) by @silox on Wed Dec 10 2014 at 17:52._
_Closed by @bmesuere on Thu Mar 26 2015 at 20:38._

**Comment by bmesuere** [comment]
The origin of this bug is in the invalidation step. If a taxon is invalidated, its complete upstream lineage is temporarily set to invalid (`-1`). The reason for this is to also invalidate any gaps between the invalid taxon **and the next valid one** in the lineage. If we don't do this, something like _undefined bacteria_ with only its superkingdom set to bacteria would still have a huge impact on the results.

The problem is that all gaps in the lineage are invalidated, and not only till the next valid taxon. For example:

```
+----------+--------------+---------+------------+-------------+--------+-----------+------------+--------+----------+------------+------------+--------+----------+------------+-----------+-------------+--------+-----------+-------+----------+-------+----------+---------------+------------------+---------+------------+----------+-------+
| taxon_id | superkingdom | kingdom | subkingdom | superphylum | phylum | subphylum | superclass | class  | subclass | infraclass | superorder | order  | suborder | infraorder | parvorder | superfamily | family | subfamily | tribe | subtribe | genus | subgenus | species_group | species_subgroup | species | subspecies | varietas | forma |
+----------+--------------+---------+------------+-------------+--------+-----------+------------+--------+----------+------------+------------+--------+----------+------------+-----------+-------------+--------+-----------+-------+----------+-------+----------+---------------+------------------+---------+------------+----------+-------+
|   944564 |            2 |      -1 |         -1 |          -1 |   1239 |        -1 |         -1 | 909932 |       -1 |         -1 |         -1 | 909929 |       -1 |         -1 |        -1 |          -1 |  31977 |        -1 |    -1 |       -1 | 29465 |       -1 |            -1 |               -1 | -671229 |         -1 |       -1 |    -1 |
+----------+--------------+---------+------------+-------------+--------+-----------+------------+--------+----------+------------+------------+--------+----------+------------+-----------+-------------+--------+-----------+-------+----------+-------+----------+---------------+------------------+---------+------------+----------+-------+
```

should be

```
+----------+--------------+---------+------------+-------------+--------+-----------+------------+--------+----------+------------+------------+--------+----------+------------+-----------+-------------+--------+-----------+-------+----------+-------+----------+---------------+------------------+---------+------------+----------+-------+
| taxon_id | superkingdom | kingdom | subkingdom | superphylum | phylum | subphylum | superclass | class  | subclass | infraclass | superorder | order  | suborder | infraorder | parvorder | superfamily | family | subfamily | tribe | subtribe | genus | subgenus | species_group | species_subgroup | species | subspecies | varietas | forma |
+----------+--------------+---------+------------+-------------+--------+-----------+------------+--------+----------+------------+------------+--------+----------+------------+-----------+-------------+--------+-----------+-------+----------+-------+----------+---------------+------------------+---------+------------+----------+-------+
|   944564 |            2 |   NULL  |       NULL |        NULL |   1239 |      NULL |       NULL | 909932 |     NULL |       NULL |       NULL | 909929 |     NULL |       NULL |      NULL |        NULL |  31977 |      NULL |  NULL |     NULL | 29465 |       -1 |            -1 |               -1 | -671229 |         -1 |       -1 |    -1 |
+----------+--------------+---------+------------+-------------+--------+-----------+------------+--------+----------+------------+------------+--------+----------+------------+-----------+-------------+--------+-----------+-------+----------+-------+----------+---------------+------------------+---------+------------+----------+-------+
```

Also note that this is an explanation for the wrong LCA, but not for the error in the tree. I'll look into that next.

@ninewise I presume you also copied this error in your rewrite?


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13503) by @bmesuere on Wed Jan 14 2015 at 14:42._

**Comment by bmesuere** [comment]
The error in the tree is related: The tree construction algorithm first adds everything from root till LCA, and then starts adding the other nodes.


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13504) by @bmesuere on Wed Jan 14 2015 at 14:46._

**Comment by bmesuere** [comment]
@ninewise I think these lines inserted before https://github.ugent.be/unipept/unipept/blob/e0e079f3ed55fff04ec7864ed743ed764d64e307/backend/java/src/main/java/org/unipept/storage/PeptideLoaderData.java#L357 will fix the problem. Can you check if this seems correct and commit it in your branch?

``` java
int rankIndex = rankIndices.get(rank);
if (lineage[rankIndex].equals("-1") && taxons.get(parentId).validTaxon) {
        for (int i = rankIndex; rankIndex > 0; i--) {
                lineage[i] = null;
        }
}
```


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13506) by @bmesuere on Wed Jan 14 2015 at 15:01._

**Comment by ninewise** [comment]
I did indeed copy this error, I'll look into it.


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13512) by @ninewise on Thu Jan 15 2015 at 12:47._

**Comment by ninewise** [comment]
I think I'll just have to modify [this](https://github.ugent.be/unipept/unipept/blob/e0e079f3ed55fff04ec7864ed743ed764d64e307/backend/java/src/main/java/org/unipept/storage/PeptideLoaderData.java#L344-L347) to mark only a part.


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13513) by @ninewise on Thu Jan 15 2015 at 12:57._

**Comment by ninewise** [comment]
Scratch that, @bmesuere approached seems better. However, is it possible that there are invalid parents (closer to the root than some valid taxon), and if so, shouldn't the missing ranks above this be ranked as invalid, too?


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13514) by @ninewise on Thu Jan 15 2015 at 13:09._

**Comment by ninewise** [comment]
This seems to be fixed:

```
$ diff <(zcat lineages.tsv.gz) <(zcat lineages_old.tsv.gz) | head -22
30c30
< 30    27  2   -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -1  -27 -1  -1  -1
---
> 30    27  2   \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  \N  -27 -1  -1  -1
129c129
< 129   232 2   -1  -1  -1  1224    -1  -1  1236    -1  -1  -1  135622  -1  -1  -1  -1  72275-1 -1  -1  226 -1  -1  -1  -232    -1  -1  -1
---
> 129   232 2   \N  \N  \N  1224    \N  \N  1236    \N  \N  \N  135622  \N  \N  \N  \N  72275\N \N  \N  226 \N  \N  \N  -232    -1  -1  -1
141,144c141,144
< 141   239 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -239    -1  -1  -1
< 142   240 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -240    -1  -1  -1
< 143   241 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -241    -1  -1  -1
< 144   242 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -242    -1  -1  -1
---
> 141   239 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 \N  \N  \N  -239    -1  -1  -1
> 142   240 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 \N  \N  \N  -240    -1  -1  -1
> 143   241 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 \N  \N  \N  -241    -1  -1  -1
> 144   242 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 \N  \N  \N  -242    -1  -1  -1
162c162
< 162   261 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -261    -1  -1  -1
---
> 162   261 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 \N  \N  \N  -261    -1  -1  -1
```


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13580) by @ninewise on Tue Jan 20 2015 at 16:26._

**Comment by bmesuere** [comment]
@ninewise nope sorry, there's still a small error. Everything up until the first valid taxon should be -1:

```
129   232 2   \N  \N  \N  1224    \N  \N  1236    \N  \N  \N  135622  \N  \N  \N  \N  72275\N \N  \N  226 -1  -1  -1  -232    -1  -1  -1
```


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13588) by @bmesuere on Tue Jan 20 2015 at 19:40._

**Comment by ninewise** [comment]
Got it:

```
 $ diff <(zcat lineages2.tsv.gz) <(zcat lineages.tsv.gz) | head -23
129c129
< 129   232 2   -1  -1  -1  1224    -1  -1  1236    -1  -1  -1  135622  -1  -1  -1  -1  72275-1 -1  -1  226 -1  -1  -1  -232    -1  -1  -1
---
> 129   232 2   \N  \N  \N  1224    \N  \N  1236    \N  \N  \N  135622  \N  \N  \N  \N  72275\N \N  \N  226 -1  -1  -1  -232    -1  -1  -1
141,144c141,144
< 141   239 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -239    -1  -1  -1
< 142   240 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -240    -1  -1  -1
< 143   241 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -241    -1  -1  -1
< 144   242 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -242    -1  -1  -1
---
> 141   239 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 -1  -1  -1  -239    -1  -1  -1
> 142   240 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 -1  -1  -1  -240    -1  -1  -1
> 143   241 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 -1  -1  -1  -241    -1  -1  -1
> 144   242 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 -1  -1  -1  -242    -1  -1  -1
162c162
< 162   261 2   -1  -1  68336   976 -1  -1  117743  -1  -1  -1  200644  -1  -1  -1  -1  49546-1 -1  -1  237 -1  -1  -1  -261    -1  -1  -1
---
> 162   261 2   \N  \N  68336   976 \N  \N  117743  \N  \N  \N  200644  \N  \N  \N  \N  49546\N \N  \N  237 -1  -1  -1  -261    -1  -1  -1
180c180
< 180   275 2   -1  -1  -1  1297    -1  -1  188787  -1  -1  -1  68933   -1  -1  -1  -1  188786-1    -1  -1  270 -1  -1  -1  -275    -1  -1  -1
---
> 180   275 2   \N  \N  \N  1297    \N  \N  188787  \N  \N  \N  68933   \N  \N  \N  \N  188786\N    \N  \N  270 -1  -1  -1  -275    -1  -1  -1
```


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13590) by @ninewise on Tue Jan 20 2015 at 22:57._

**Comment by bmesuere** [comment]
:+1: 


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-13597) by @bmesuere on Wed Jan 21 2015 at 11:40._

**Comment by bmesuere** [comment]
fixed as of 2.5.2


_[Original comment](https://github.ugent.be/unipept/unipept/issues/443#issuecomment-15226) by @bmesuere on Thu Mar 26 2015 at 20:38._
